src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -9,121 +9,173 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine
-
-                    // Initialize the proxy.
-                    operationManagerProxy.Initialize(this.skipDefaultAdapters);
-
-                    // Start the test host associated to the proxy.
-                    operationManagerProxy.SetupChannel(
-                        criteria.Sources,
-                        criteria.RunSettings,
-                        eventsHandler);
+                    if (!res.Result)
+                    {
+                        this.proxySetupFailed = true;
+                    }
                 });
             }
 
             // Wait for proxy creation to be over.
             Task.WaitAll(taskList);
 
+            // Dispose of all proxies if even one of them failed during setup.
+            if (this.proxySetupFailed)
+            {
+                this.DisposeProxies();
+                return false;
+            }
+
             // Make the session available.
-            TestSessionPool.Instance.AddSession(testSessionInfo, this);
+            if (!TestSessionPool.Instance.AddSession(this.testSessionInfo, this))
+            {
+                this.DisposeProxies();
+                return false;
+            }
 
             // Let the caller know the session has been created.
-            eventsHandler.HandleStartTestSessionComplete(testSessionInfo);
+            eventsHandler.HandleStartTestSessionComplete(this.testSessionInfo);
+            return true;
         }
 
         /// <inheritdoc/>
-        public void StopSession()
+        public virtual bool StopSession()
         {
-            // TODO (copoiena): Do nothing for now because in the current implementation the
-            // testhosts are disposed of right after the test run is done. However, when we'll
-            // decide to re-use the testhosts for discovery & execution we'll perform some
-            // changes for keeping them alive indefinetely, so the responsability for killing
-            // testhosts will be with the users of the vstest.console wrapper. Then we'll need
-            // to be able to dispose of the testhosts here.
-
-            // foreach (var kvp in this.proxyMap)
-            // {
-            // }
+            lock (this.lockObject)
+            {
+                if (this.testSessionInfo == null)
+                {
+                    return false;
+                }
+                this.testSessionInfo = null;
+            }
+
+            this.DisposeProxies();
+            return true;
         }
 
         /// <summary>
         /// Dequeues a proxy to be used either by discovery or execution.
         /// </summary>
         /// 
+        /// <param name="source">The source to be associated to this proxy.</param>
+        /// <param name="runSettings">The run settings.</param>
+        /// 
         /// <returns>The dequeued proxy.</returns>
-        public ProxyOperationManager DequeueProxy()
+        public virtual ProxyOperationManager DequeueProxy(string source, string runSettings)
```

Comment:
For shared proxies we map each source to the same `ProxyOperationManager` in the internal mapping held by the `ProxyTestSessionManager`. As such we can, in theory, deque the same proxy for each of the sources provided the proxy is available in the first place (i.e. not taken by some other discovery/run operation). In practice though, when running on shared hosts, the test discovery/run call will end up getting all the sources instead of individual sources managed by a `ParallelProxyDiscoveryManager`/`ParallelProxyExecutionManager` as it happens with non-shared testhosts, and so we will only deque the proxy once for all the sources, run test discovery/execution on all of them and then return the proxy back to the pool.